fastly/errors.go

```diff
@@ -9,180 +9,44 @@ import (
-var ErrMissingWAFVersionID = errors.New("missing required field 'WAFVersionID'")
-
-// ErrMissingWAFActiveRuleList is an error that is returned when an input struct
-// requires a list of WAF active rules, but it is empty.
-var ErrMissingWAFActiveRuleList = errors.New("WAF active rules slice is empty")
-
-// ErrMissingWAFRuleExclusionNumber is an error that is returned when an input struct
-// requires a "WAFExclusionNumber" key, but one was not set.
-var ErrMissingWAFRuleExclusionNumber = errors.New("missing required field 'WAFExclusionNumber'")
-
-// ErrMissingWAFRuleExclusion is an error that is returned when an input struct
-// requires a "WAFRuleExclusion" key, but one was not set.
-var ErrMissingWAFRuleExclusion = errors.New("missing required field 'WAFRuleExclusion'")
-
-// ErrMissingOWASPID is an error that is returned was an input struct
-// requires a "OWASPID" key, but one was not set
-var ErrMissingOWASPID = errors.New("missing required field 'OWASPID'")
-
-// ErrMissingRuleID is an error that is returned was an input struct
-// requires a "RuleID" key, but one was not set
-var ErrMissingRuleID = errors.New("missing required field 'RuleID'")
-
-// ErrMissingConfigSetID is an error that is returned was an input struct
-// requires a "ConfigSetID" key, but one was not set
-var ErrMissingConfigSetID = errors.New("missing required field 'ConfigSetID'")
-
-// ErrMissingWAFList is an error that is returned was an input struct
-// requires a list of WAF id's, but it is empty
-var ErrMissingWAFList = errors.New("WAF slice is empty")
-
-// ErrMissingPool is an error that is returned when an input struct requires
-// a "Pool" key, but one was not set.
-var ErrMissingPool = errors.New("missing required field 'Pool'")
-
-// ErrMissingServer is an error that is returned when an input struct requires
-// a "Server" key, but one was not set.
-var ErrMissingServer = errors.New("missing required field 'Server'")
-
-// ErrMissingAddress is an error that is returned when an input struct requires
-// a "Address" key, but one was not set.
-var ErrMissingAddress = errors.New("missing required field 'Address'")
-
-// ErrBatchUpdateMaximumItemsExceeded is an error that indicates that too many batch operations are being executed.
-// The Fastly API specifies an maximum limit.
-var ErrBatchUpdateMaximumOperationsExceeded = errors.New("batch modify maximum operations exceeded")
-
-// ErrMissingKMSKeyID is an error that is returned from an input struct that requires
-// a "ServerSideEncryptionKMSKeyID" key, but one was not set.
-var ErrMissingKMSKeyID = errors.New("missing required field 'ServerSideEncryptionKMSKeyID'")
-
-// ErrMissingCertBlob is an error that is returned from an input struct that requires
-// a "CertBlob" key, but one was not set.
-var ErrMissingCertBlob = errors.New("missing required field 'CertBlob'")
+// FieldError represents a custom error type for API data fields.
+type FieldError struct {
+	kind    string
+	message string
+}
```

Comment:
While I agree that a new custom error type `FieldError` is a good and needed abstraction to the problem; I'm slightly confused as to why you've removed the package level constants to store the unique instances of the errors? 

We now have 100+ instances of `NewFieldError("ServiceID")` across the codebase instead of a single constant definition in this file which can then be referenced elsewhere (and reduce the memory allocation significantly). The bigger issue is we then are forced to assert against the raw error string in every test, meaning we now have:
```diff
-      if err != ErrMissingServiceVersion {
+      if err.Error() != "missing required field 'ServiceID'" {
```
 in many places which I find very brittle compared to the previous implementation. I.e. if we chose to change the implementation of `FieldError` to return a slightly different message at a later date all of these tests will instantly fail.

Therefore, I'd suggest we go with `FieldError` but keep the constants and their references which should dramatically reduce the change set of this PR. Alternatively if you want to improve the error assertions accross the codebase, you could use the fancy new [`error.Is()`](https://golang.org/pkg/errors/#Is) introduced in Go 1.13.

Unless I'm missing something and there was a good reason you when with this pattern of a new instance of `FieldError` per file?